### PR TITLE
[INTEL MKL] Enabling TF + OneDNN in stock TF build

### PR DESCRIPTION
**Pull request by gzmkl**
Refactor code to do the following:

1) code that is included in stock TF build will be guarded by macro INTEL_MKL

2) code that is included only when --config=mkl is used is now guarded by macro ENABLE_MKL

3) when building stock TF for x86 Linux/Windows the macro INTEL_MKL will be defined, and won't be defined for other architectures.

4) Added runtime env variable ENABLE_ONEDNN_OPTS that will enable oneDNN optimizations in stock TF.

5) for config=mkl, ENABLE_MKL will be defined.

This PR assumes 

a) oneDNN is upgraded to 2.1 (PR #: https://github.com/tensorflow/tensorflow/pull/47743) and 

b) stock TF and TF+oneDNN are using same oneDNN build (https://github.com/tensorflow/tensorflow/pull/47679)

### CODE REVIEW

**Comment by gbaned** [comment]
@gzmkl  Can you please resolve conflicts? Thanks!

**Comment by gzmkl** [comment]
@gbaned  The merge conflicts were related to oneDNN upgrade PR. They are gone after the upgrade PR is merged

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1764]: Do we still need this `#ifdef INTEL_MKL`? Can the whole block just be:
```c++
  if (IsMKLEnabled()) {
    return is_batch_norm_candidate() || is_batch_norm_fusion_candidate() ||
           IsContractionWithAdd(ctx, node_index);
  }
  return is_relu_biasadd_conv2d_candidate() || is_batch_norm_candidate() ||
         is_batch_norm_fusion_candidate();
```
?
(We don't need `(void)is_relu_biasadd_conv2d_candidate;` anymore because the variable is now used.)

If you prefer to keep `#ifdef INTEL_MKL` and clean up later, then I think this should be fine:
```c++
#ifdef INTEL_MKL
  (void)is_relu_biasadd_conv2d_candidate;  // To fix unused variable error.
  if (IsMKLEnabled()) {
    return is_batch_norm_candidate() || is_batch_norm_fusion_candidate() ||
           IsContractionWithAdd(ctx, node_index);
  }
#endif  // INTEL_MKL
  return is_relu_biasadd_conv2d_candidate() || is_batch_norm_candidate() ||
         is_batch_norm_fusion_candidate();
```

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1890]: Shouldn't this be `is_onednn_enabled`? I.e., we should only try this rewrite when oneDNN is enabled.
```suggestion
    if (is_onednn_enabled && allow_non_differentiable_rewrites &&
```

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:1901]: Same question here.
```suggestion
    if (is_onednn_enabled && allow_non_differentiable_rewrites &&
```

File: tensorflow/core/grappler/optimizers/remapper_test.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:467]: Nit:
```suggestion
#endif  // !defined(ENABLE_MKL)
```

File: tensorflow/core/grappler/optimizers/remapper_test.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:464]: Does `bfloat16` only work with non-native mode right now? If so, please add a comment saying so. Same as other places in this file.

File: tensorflow/core/grappler/optimizers/remapper_test.cc
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:748]: Nit: Same here.
```suggestion
#endif  // !defined(ENABLE_MKL)
```

If this is the formatter's doing, please feel free to leave it like this.

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc:245]: Please remove this. The whole file is already under `#ifdef INTEL_MKL` (see line 26).

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc:1692]: Nit:
```suggestion
#endif  // INTEL_MKL && ENABLE_MKL
```

File: tensorflow/core/util/util.cc
Comment (penpornk) [inline tensorflow/core/util/util.cc:125]: Can we move `#ifdef INTEL_MKL` inside this function? E.g.,
```c++
bool IsMKLEnabled() {
#ifndef INTEL_MKL
  return false;
#else 
  ...
#endif  // !INTEL_MKL
}
```

I'm also fine if you want to keep the whole function inside `#ifdef INTEL_MKL` for now. (E.g., `IsMKLEnabled` won't exist if `-DINTEL_MKL` is not provided.)

File: tensorflow/tensorflow.bzl
Comment (penpornk) [inline tensorflow/tensorflow.bzl:367]:
```suggestion
        # Enable additional ops (e.g., ops with non-NHWC data layout) and optimizations for Intel builds using
```
(Please also re-wrap the line.)

File: third_party/mkl/build_defs.bzl
Comment (penpornk) [inline third_party/mkl/build_defs.bzl:92]: Nit
```suggestion
      depending on the platform. x86 Linux/Windows with or without --config=mkl will
```
(Please also re-wrap the line.)

File: third_party/mkl/build_defs.bzl
Comment (penpornk) [inline third_party/mkl/build_defs.bzl:30]: Should we delete this one?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite_test.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite_test.cc:16]: We had difficulty in refactoring all mkl related tests at run-time detect if MKL is enabled at run-time. 
We plan to refactor and update later. Currently these tests run only when ENABLE_MKL is true.

Comment (penpornk) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite_test.cc:16]: @gzmkl Thank you for the explanation!

File: third_party/mkl/build_defs.bzl
Comment (gzmkl) [reply third_party/mkl/build_defs.bzl:30]: We think that ARM may need it when compiling with oneDNN. We will probably remove it later with refactoring.

**Comment by penpornk** [comment]
@gzmkl By the way, if any change requires long testing time and you prefer to do it later. Please feel free to just add a TODO instead.

File: third_party/mkl/build_defs.bzl
Comment (penpornk) [reply third_party/mkl/build_defs.bzl:30]: Got it. Could you please add a comment explaining this in the code as well? Thank you!

File: tensorflow/core/common_runtime/threadpool_device.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/threadpool_device.cc:35]: Yes, we are fixing it to make code same.

File: tensorflow/core/common_runtime/mkl_tfconversion_pass.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/mkl_tfconversion_pass.cc:16]: Good suggestion, code change is coming.

File: tensorflow/core/common_runtime/process_util.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/process_util.cc:104]: Yes, it should be ENABLE_MKL. Comment should also be update. Fix is coming...

File: tensorflow/core/common_runtime/threadpool_device.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/threadpool_device.cc:39]: Yes, we are doing the change. Also replace INTEL_MKL with ENABLE_MKL due to line 35 change.

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc
Comment (gzmkl) [reply tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc:1692]: This type will be fixed. change is coming...

File: tensorflow/core/util/util.cc
Comment (gzmkl) [reply tensorflow/core/util/util.cc:125]: Actually your suggestion was my original implementation.  I will make the change back even though IsMKLEnabled() 
always exists. This will make TF code much cleanly -- currently wherever IsMKLEnabled() is invoked, we have to add 
"#ifdef INTEL_MKL" protection :). 

For this PR, I will change IsMKLEnable method. We will refactor its usage places later.

File: tensorflow/tensorflow.bzl
Comment (gzmkl) [reply tensorflow/tensorflow.bzl:367]: Will make code change per suggestion

File: third_party/mkl/build_defs.bzl
Comment (gzmkl) [reply third_party/mkl/build_defs.bzl:92]: Yes, will fix with wrap.

File: third_party/mkl/build_defs.bzl
Comment (gzmkl) [reply third_party/mkl/build_defs.bzl:30]: Comment will be added.

**Comment by gzmkl** [comment]
@penpornk  Thank you for the code review. Most code change suggestions are very good and I have done code change. 
Local test will take a while but will try to push them to the PR branch today.

**Comment by penpornk** [comment]
@gzmkl Thank you very much for the quick responses! Please take your time. :) 
The oneDNN v2.1 upgrade is still not in the clear. It made one test timed out (due to long compilation time) so I submitted a [fix](https://github.com/tensorflow/tensorflow/commit/bf4516eec17a6685ef07ed823529dce140b842c2) this morning. I'll have to wait for tonight's nightly test results to make sure v2.1 will stick. The soonest this PR can be merged is tomorrow (if nothing else goes wrong).

**Comment by gzmkl** [comment]
@penpornk I have committed changes per your suggestions. Please let me know if there is anything missing.
Thanks a lot.

File: tensorflow/core/common_runtime/threadpool_device.cc
Comment (penpornk) [reply tensorflow/core/common_runtime/threadpool_device.cc:39]: The comment should be added to line 39 though. `#endif` in line 38 is for `#ifdef _OPENMP`. I will change this internally.

File: tensorflow/core/util/util.cc
Comment (penpornk) [reply tensorflow/core/util/util.cc:124]: Sounds good. Thank you! :)

**Review by penpornk — APPROVED** [review]
Thank you for the quick response! Several of my comments are still not addressed. But I can just make the rest of the minor modifications (and add TODOs for refactors) myself. I'm going to pull the PR in to test now.

There are two comments that I need answers though. Could you please reply to these questions? [Q1](https://github.com/tensorflow/tensorflow/pull/47745#discussion_r597397968), [Q2](https://github.com/tensorflow/tensorflow/pull/47745#discussion_r597398023)

**Comment by gbaned** [comment]
@gzmkl  Can you please check @penpornk's comments and keep us posted ? Thanks!

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (gzmkl) [reply tensorflow/core/grappler/optimizers/remapper.cc:1901]: We do not support contraction with fused batch norm for now.  We will have a PR (https://github.com/tensorflow/tensorflow/pull/47828) to support this case. 

So the code is logically correct.

File: tensorflow/core/common_runtime/threadpool_device.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/threadpool_device.cc]: Fixed!

**Comment by gzmkl** [comment]
> @gzmkl Can you please check @penpornk's comments and keep us posted ? Thanks!

Yes, I just did that. Thanks!

**Comment by penpornk** [comment]
@gzmkl The changes from this PR have be merged in https://github.com/tensorflow/tensorflow/commit/4a24610212cb69f1bc36cfaff19af064ec35a1d6. Somehow Github doesn't mark it as merged. I'm closing this PR now. Thank you again for the PR! :)

**Comment by ScottTodd** [comment]
Hey, I'm noticing some build failures in a downstream project ([IREE](https://github.com/google/iree)) after this was merged. Full logs are [here](https://source.cloud.google.com/results/invocations/8747de95-2981-4fa3-bc1d-591a5f868a88/targets/iree%2Fgcp_ubuntu%2Fcmake-bazel%2Flinux%2Fx86-swiftshader%2Fpresubmit/log). Maybe we're using different flags/compilers to build.

Relevant logs snippet:

```
ERROR: /home/kbuilder/.cache/bazel/_bazel_kbuilder/1900d0fac5123d725c8d2e08b3f8c209/external/org_tensorflow/tensorflow/core/kernels/mkl/BUILD:299:22: C++ compilation of rule '@org_tensorflow//tensorflow/core/kernels/mkl:mkl_softmax_op' failed (Exit 1): clang failed: error executing command /usr/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 262 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command /usr/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 262 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from external/org_tensorflow/tensorflow/core/kernels/mkl/mkl_softmax_op.cc:22:
In file included from external/org_tensorflow/tensorflow/core/framework/numeric_op.h:19:
In file included from external/org_tensorflow/tensorflow/core/framework/op_kernel.h:24:
In file included from external/org_tensorflow/tensorflow/core/framework/allocator.h:28:
In file included from external/org_tensorflow/tensorflow/core/platform/logging.h:27:
external/org_tensorflow/tensorflow/core/platform/default/logging.h:280:9: error: call to function 'operator<<' that is neither visible in the template definition nor found by argument-dependent lookup
  (*os) << v;
        ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:339:3: note: in instantiation of function template specialization 'tensorflow::internal::MakeCheckOpValueString<dnnl::memory::format_tag>' requested here
  MakeCheckOpValueString(comb.ForVar1(), v1);
  ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:384:1: note: in instantiation of function template specialization 'tensorflow::internal::MakeCheckOpString<dnnl::memory::format_tag, dnnl::memory::format_tag>' requested here
TF_DEFINE_CHECK_OP_IMPL(Check_NE, !=)  // Use CHECK(x == NULL) instead.
^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:357:38: note: expanded from macro 'TF_DEFINE_CHECK_OP_IMPL'
      return ::tensorflow::internal::MakeCheckOpString(v1, v2, exprtext); \
                                     ^
external/org_tensorflow/tensorflow/core/util/mkl_util.h:472:7: note: in instantiation of function template specialization 'tensorflow::internal::Check_NEImpl<dnnl::memory::format_tag, dnnl::memory::format_tag>' requested here
      DCHECK_NE(format_tag, memory::format_tag::undef);
      ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:418:31: note: expanded from macro 'DCHECK_NE'
#define DCHECK_NE(val1, val2) CHECK_NE(val1, val2)
                              ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:405:30: note: expanded from macro 'CHECK_NE'
#define CHECK_NE(val1, val2) CHECK_OP(Check_NE, !=, val1, val2)
                             ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:401:40: note: expanded from macro 'CHECK_OP'
#define CHECK_OP(name, op, val1, val2) CHECK_OP_LOG(name, op, val1, val2)
                                       ^
external/org_tensorflow/tensorflow/core/platform/default/logging.h:395:31: note: expanded from macro 'CHECK_OP_LOG'
      ::tensorflow::internal::name##Impl(                      \
                              ^
<scratch space>:124:1: note: expanded from here
Check_NEImpl
^
external/org_tensorflow/tensorflow/core/util/mkl_util.h:184:22: note: 'operator<<' should be declared prior to the call site or in namespace 'dnnl'
inline std::ostream& operator<<(std::ostream& os,
                     ^
1 error generated.
```

This is the problematic line:
https://github.com/tensorflow/tensorflow/blob/a45aa6f1731cc7e89b01b0d0da199841f28213c8/tensorflow/core/util/mkl_util.h#L472

This can be fixed by either removing that line or moving this `operator<<` from `namespace tensorflow` to `namespace dnnl`: https://github.com/tensorflow/tensorflow/blob/a45aa6f1731cc7e89b01b0d0da199841f28213c8/tensorflow/core/util/mkl_util.h#L184-L194

How do you want to proceed?

**Comment by penpornk** [comment]
@ScottTodd Sorry about the break and thank you for proposing the fix!
I'll leave this to @gzmkl and @agramesh1 to decide. I think making changes in TensorFlow is probably easier than in oneDNN (namespace `dnnl` is in an external library.)

**Comment by agramesh1** [comment]
@penpornk thanks, we are looking at it now.

**Comment by ScottTodd** [comment]
> @ScottTodd Sorry about the break and thank you for proposing the fix!
> I'll leave this to @gzmkl and @agramesh1 to decide. I think making changes in TensorFlow is probably easier than in oneDNN (namespace `dnnl` is in an external library.)

Changes don't need to be made in the external library, simply putting

```
namespace dnnl {
inline std::ostream& operator<<(std::ostream& os,
                                const memory::format_tag& tag) {
  if (tag == memory::format_tag::undef) {
    os << "undef";
  } else if (tag == memory::format_tag::any) {
    os << "any";
  } else {
    os << "invalid";
  }
  return os;
}
}  // namespace
```

above `namespace tensorflow`: https://github.com/tensorflow/tensorflow/blob/a45aa6f1731cc7e89b01b0d0da199841f28213c8/tensorflow/core/util/mkl_util.h#L52-L56

works when I build locally. (I think `memory::format_tag` is inlined without being fully qualified as `dnnl::memory::format_tag`, the macros aren't compiling as expected)

That's just one solution though. Not sure what works best for your projects.

**Comment by gzmkl** [comment]
After discussing with team and reviewing the related code in mkl_util.h,  removing the related line of code is the right choice.
We will submit an PR.

**Comment by gzmkl** [comment]
@ScottTodd @penpornk  A PR to fix the issue has been submitted here https://github.com/tensorflow/tensorflow/pull/48030.
Thanks!

**Pull request closed without merging**
